### ISSUE TITLE
MethodError from @reaction_network with parameters

### ISSUE THREAD

**Issue by improbable22** [JuliaDiffEq/DifferentialEquations.jl#269]
I was trying out the `@reaction_network` macro from [the manual](http://docs.juliadiffeq.org/latest/models/biological.html), and while the first few examples work, the ones with parameters do not. The errors I get are as follows: 

```
julia> Pkg.status("DifferentialEquations")
 - DifferentialEquations         4.1.0

julia> rn = @reaction_network rType begin
           (kB, kD), X + Y ↔ XY
       end kB, kD
ERROR: MethodError: DiffEqBiological.@reaction_network(::Symbol, ::Expr, ::Expr) is ambiguous. Candidates:
  @reaction_network(name::ANY, scale_noise::ANY, ex::Expr, p...) in DiffEqBiological at /Users/me/.julia/v0.6/DiffEqBiological/src/reaction_network.jl:58
  @reaction_network(name::ANY, ex::Expr, p...) in DiffEqBiological at /Users/me/.julia/v0.6/DiffEqBiological/src/reaction_network.jl:53
Possible fix, define
  @reaction_network(::ANY, ::Expr, ::Expr, ::Vararg{Any,N} where N)

julia> rs = @reaction_network begin
         c1, X --> 2X
         c2, X --> 0
         c3, 0 --> X
       end c1 c2 c3
ERROR: MethodError: no method matching @reaction_network(::Expr, ::Symbol, ::Symbol, ::Symbol)
Closest candidates are:
  @reaction_network(::ANY, ::Expr, ::Any...) at /Users/me/.julia/v0.6/DiffEqBiological/src/reaction_network.jl:53
  @reaction_network(::ANY, ::ANY, ::Expr, ::Any...) at /Users/me/.julia/v0.6/DiffEqBiological/src/reaction_network.jl:58
  @reaction_network(::Expr) at /Users/me/.julia/v0.6/DiffEqBiological/src/reaction_network.jl:64
```

**Comment by ChrisRackauckas** [comment]
It's not parameters, it's two-way reactions. Xref: https://github.com/JuliaDiffEq/DiffEqBiological.jl/issues/30

**Comment by improbable22** [comment]
Ah so it is, thanks! 

My second example can be fixed by including `rType`:

```
rs = @reaction_network rType begin
                c1, X --> 2X
                c2, X --> 0
                c3, 0 --> X
              end c1 c2 c3
```

**Comment by improbable22** [comment]
I still get an error from the [last example](http://docs.juliadiffeq.org/latest/models/biological.html#Example:-Michaelis-Menten-Enzyme-Kinetics-1) after adding `rType`:

```
julia> rn = @reaction_network rType begin
           c1, S + E --> SE               
           c2, SE --> S + E 
           c3, SE --> P + E
       end c1 c2 c3
ERROR: ArgumentError: Second argument must be of Symbol type

julia> E
ERROR: UndefVarError: E not defined
```
After messing around a bit, it seems that the letter `E` is the problem here... replacing it with `EE` fixes it. Seems bizarre, if easy to work around! 
```
julia> rn = @reaction_network rType begin
           c1, S + EE --> SE               
           c2, SE --> S + EE 
           c3, SE --> P + EE
       end c1 c2 c3
rType(#690, Expr[:(-1 * c1 * S * EE + c2 * SE), ... [clipped]
```

**Comment by Gaussia** [comment]
More or less yes. I actually have a solution mostly implemented on my computer, but then it became weekend and I got busy. I have tracked the problem down to the line:
´´´julia
symjac = Expr(:quote, calculate_jac(f_rhs, syms))
´´´
I am not really sure about this part, (@korsbo did the symbolic stuff). However the ´´´calculate_jac´´´ function cannot handle certain input. I did some tests and I think that 'e', 'E' and 'I' all give problems, but that should be the only issues (I think 'π' did work. Might be others but I tested all letters).

Right now my solution is simply to add a check which detect if one of these are used as reactant names, and in that case gives an error explaining the problem (same if 't' is used).

The first part I will have a check at tomorrow.

**Comment by ChrisRackauckas** [comment]
IMO it might be worthwhile to drop the Jacobian calculations and just hold off on that until ModelingToolkit.jl is used as the IR.

**Comment by korsbo** [comment]
Agreed. The Jacobian is currently calculated, but not actually used. If we are planning to move such pre-processing to a separate package, then we should just drop it here.

**Comment by Gaussia** [comment]
Removing it until it is actually used might be the best idea, especially if we might not actually use it at all.

Moving on to the issues @improbable22  have:
```julia
rn = @reaction_network rType begin
           (kB, kD), X + Y ↔ XY
end kB, kD
```
The problem here is that the parameters are separated by a `,`. In that example thie `,` have been incorrectly used. In the next example the `,` is correctly not used.
```julia
rs = @reaction_network begin
         c1, X --> 2X
         c2, X --> 0
         c3, 0 --> X
end c1 c2 c3
```
Your correctly identifies that adding a type fixes this error and a type is right now required. If you had not added parameters you would have gotten an error message that you were using the old syntax and that the DSL had been overhauled. In the long run though when that might not be needed I'd like to create a dispatch for when type is not given (for when it is not needed) and some default type is instead assigned.

... and the last problem has to do with what ´´´symjac = Expr(:quote, calculate_jac(f_rhs, syms))´´´ accept and doesn't accept.
